ftw/referencewidget/browser/jsongenerator.py

```diff
@@ -0,0 +1,76 @@
+from ftw.referencewidget.browser.utils import get_selectable_types
+from ftw.referencewidget.browser.utils import get_traversal_types
+from Products.CMFCore.utils import getToolByName
+from Products.Five import BrowserView
+import json
+
+
+class ReferenceJsonEndpoint(BrowserView):
+
+    def __call__(self):
+        batchsize = 20
+        page = 1
+        widget = self.context
+        effective_path = ""
+        if widget.request.get('page'):
+            page = int(widget.request.get('page'))
+        if widget.request.get('start'):
+            effective_path = widget.request.get('start')
+        elif not widget.start:
+            effective_path = '/'.join(widget.form.context.getPhysicalPath())
+        else:
+            import pdb; pdb.set_trace()
+            if not callable(widget.start):
+                effective_path = widget.start
+            else:
+                effective_path = widget.start()
+
+        current_depth = len(effective_path.split('/'))
+        traversel_type = get_traversal_types(widget)
+        query = {'portal_type': traversel_type,
```

Comment:
Please use decent methods, so that it can be subclassed when needed without having to fork the complete method. This method is too long.